File: 57-insert-interval/57-insert-interval.scala
object Solution {
    def insert(
        intervals: Array[Array[Int]],
        newInterval: Array[Int]
    ): Array[Array[Int]] = {
        val lower = findLower(intervals, newInterval)
        val higher = findHigher(intervals, newInterval)
        var res: Array[Array[Int]] = Array()

        res = res ++: intervals.slice(0, lower + 1)

        for (i <- lower + 1 to higher - 1) {
            val intv = intervals(i)
            newInterval(0) = Math.min(intv(0), newInterval(0))
            newInterval(1) = Math.max(intv(1), newInterval(1))
        }
        
        res = res :+ newInterval
        res = res ++: intervals.slice(higher, intervals.length)

        res
    }

    def findLower(intvs: Array[Array[Int]], nIntv: Array[Int]): Int = {
        var start = 0
        var end = intvs.length - 1

        while (start <= end) {
            val mid = (start + end) / 2
            val intv = intvs(mid)
            if (intv(1) < nIntv(0)) start = mid + 1
            else end = mid - 1
        }

        end
    }

    def findHigher(intvs: Array[Array[Int]], nIntv: Array[Int]): Int = {
        var start = 0
        var end = intvs.length - 1

        while (start <= end) {
            val mid = (start + end) / 2
            val intv = intvs(mid)
            if (intv(0) <= nIntv(1)) start = mid + 1
            else end = mid - 1
        }

        start
    }
}

File: 57-insert-interval/57-insert-interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> result = new ArrayList<>();
        
        int[] current = newInterval;
        
        for(int i = 0; i < intervals.length; i++) {
            int[] interval = intervals[i];
            
            if (current == null) {
                result.add(interval);
                continue;
            }
            
            if (isOverlapping(interval, current)) {
                current[0] = Math.min(interval[0], current[0]);
                current[1] = Math.max(interval[1], current[1]);
                continue;
            }
            
            
            if (interval[1] < current[0]) {
                result.add(interval);
                continue;
            }
            
            result.add(current);
            current = null;
            result.add(interval);
        }

        if (current != null) {
            result.add(current);
        }
        
        return result.toArray(int[][]::new);
    }
    
    boolean isOverlapping(int[] interval1, int[] interval2) {
        return Math.max(interval1[0], interval2[0]) <= Math.min(interval1[1], interval2[1]);
    }
}